src/tainting/Taint_shape.mli

```diff
@@ -0,0 +1,89 @@
+(** Taint shapes *)
+
+module Fields : Map.S with type key = Taint.offset
+
+(**
+ * For example, a record expression #{ a: "tainted", b: "safe" } would have
+ * the shape `obj {| a: ref<{"tainted"}>(_|_) |}`. Note that `b` is omitted
+ * because it has no taint.
+ *
+ * 'Obj' shapes are open, if a field is not specified then it has the same
+ * taint as its parent 'ref'. *)
+type shape =
+  | Bot  (** _|_, don't know or don't care *)
+  | Obj of obj  (** a struct-like thing *)
+
+and ref =
+  | Ref of Xtaint.t * shape
+      (** A "reference cell", like a `ref` in OCaml, or a variable in C.
+   *
+   * A ref may be explicitly tainted ('`Tainted'), not explicitly tainted
+   * ('`None' / "0"),  or explicitly clean ('`Clean' / "C").
+   *
+   * A ref that is not explicitly tainted inherits any taints from "parent"
+   * refs. A ref that is explicitly clean it is clean regardless.
+   *
+   * For example, given a variable `x` and the following statements:
+   *
+   *     x.a := "tainted";
+   *     x.a.u := "clean";
+   *
+   * We could assign the following shape to `x`:
+   *
+   *     ref<0>( obj {|
+   *             a: ref<"tainted">( obj {|
+   *                     u: ref<C>(_|_)
+   *                     |} )
+   *             |} )
+   *
+   * We have that `x` itself has no taint directly assigned to it, but `x.a` is
+   * tainted (by the string `"tainted"`). Other fields like `x.b` are not tainted.
+   * When it comes to `x.a`, we have that `x.a.u` has been explicitly marked clean,
+   * so `x.a.u` will be considered clean despite `x.a` being tainted. Any other field
+   * of `x.a` such as `x.a.v` will inherit the same taint as `x.a`.
```

Comment:
That's a fair suggestion! I'll probably rename it. But since it's a minor thing I'll add a TODO note for now, to avoid delaying the merge of this for much longer, this PR touches many things and I keep having to rebase.